### PR TITLE
LeakAgent

### PR DESCRIPTION
**Pull request by Protryon**

### CODE REVIEW

File: leakproto/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+description = "Metrics forwarding agent for LeakSignal"
+keywords = [ "leaksignal", "metrics", "agent" ]
+readme = "../README.md"
+homepage = "https://www.leaksignal.com/"
+
+[dependencies]
+tonic = { version = "0.8", optional = true }
+prost = { workspace = true }
+
+serde = { workspace = true, features = ["derive", "rc"] }
```
Comment (0rphon) [inline leakproto/Cargo.toml:17]: should only need serde if tonic feature is enabled

File: leakagent/src/ingestion/match_data.rs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        let request_headers = request
+            .headers
+            .into_iter()
+            .map(|x| (x.name, x.value))
+            .collect::<IndexMap<String, Option<String>>>();
+        let response_headers = response
+            .headers
+            .into_iter()
+            .map(|x| (x.name, x.value))
+            .collect::<IndexMap<String, Option<String>>>();
```
[the remark this answers is not in the document]
Comment (Protryon) [reply leakagent/src/ingestion/match_data.rs:82]: Not sure what you mean here, can you be more specific?

File: leakagent/src/ingestion/proxy_pull.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                                Err(e) => {
+                                    warn!(
+                                        "failed to open proxy file '{}' for reading: {e}",
+                                        path.display()
+                                    );
+                                    let mut response = hyper::Response::new(BoxBody::default());
+                                    *response.status_mut() = StatusCode::INTERNAL_SERVER_ERROR;
+                                    return Ok(response);
+                                }
+                            };
```
[the remark this answers is not in the document]
Comment (Protryon) [reply leakagent/src/ingestion/proxy_pull.rs:138]: Doesn't look applicable here. We are pulling the error out.

File: leakagent/src/ingestion/proxy_pull.rs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                                .extend(key.split('/'));
+                            let request = match CLIENT.get(origin).send().await {
+                                Ok(request) => request,
+                                Err(e) => {
+                                    error!("failed to initiate remote passthrough transfer: {e:?}");
+                                    let mut response = hyper::Response::new(BoxBody::default());
+                                    *response.status_mut() = StatusCode::BAD_GATEWAY;
+                                    return Ok(response);
+                                }
+                            };
```
[the remark this answers is not in the document]
Comment (Protryon) [reply leakagent/src/ingestion/proxy_pull.rs:97]: Ditto

File: leakagent/src/ingestion/match_data.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (0rphon) [reply leakagent/src/ingestion/match_data.rs:82]: oh you know what nevermind. i meant you should impl `From<Header>` for indexmap but just realized youre operating over a vec so it wouldnt work

**Review by 0rphon — APPROVED** [review]

**Pull request merged**